File: src/test/kotlin/org/paradise/ipaq/integration/IntegrationConfig.java
package org.paradise.ipaq.integration;

import au.com.auspost.signing.aws.KmsEncryption;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.connection.RedisConnection;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.data.redis.core.ValueOperations;

import java.io.IOException;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Created by terrence on 26/5/17.
 */
@Configuration
public class IntegrationConfig {

    public static final String ENCRYPTED_PERSISTED_VALUE = "Encrypted persisted value";

    @Value("${app.encryptionKey.arn}")
    private String arn;

    @Bean
    public RedisConnectionFactory connectionFactory() {

        RedisConnectionFactory redisConnectionFactory = mock(RedisConnectionFactory.class);
        RedisConnection redisConnection = mock(RedisConnection.class);

        when(redisConnectionFactory.getConnection()).thenReturn(redisConnection);

        return redisConnectionFactory;
    }

    @Bean
    public StringRedisTemplate stringRedisTemplate() {

        StringRedisTemplate stringRedisTemplate = mock(StringRedisTemplate.class);
        ValueOperations<String, String> valueOperations = mock(ValueOperations.class);

        when(stringRedisTemplate.opsForValue()).thenReturn(valueOperations);

        doNothing().when(valueOperations).set(any(String.class), any(String.class));

        when(valueOperations.get(eq(AbstractIntegrationTest.TOKEN))).thenReturn(ENCRYPTED_PERSISTED_VALUE);
        when(valueOperations.get(eq(AbstractIntegrationTest.BEARER))).thenReturn(ENCRYPTED_PERSISTED_VALUE);

        doNothing().when(stringRedisTemplate).delete(anyString());

        return stringRedisTemplate;
    }

    @Bean
    public KmsEncryption kmsEncryption() throws IOException {

        KmsEncryption kmsEncryption = mock(KmsEncryption.class);

        when(kmsEncryption.encrypt(any(String.class))).thenReturn(ENCRYPTED_PERSISTED_VALUE);

        return kmsEncryption;
    }

}

File: src/test/kotlin/org/paradise/ipaq/integration/IntegrationConfig.kt
package org.paradise.ipaq.integration

import org.apache.commons.io.IOUtils
import org.apache.commons.lang3.StringUtils
import org.mockito.Matchers.any
import org.mockito.Matchers.anyString
import org.mockito.Mockito.*
import org.paradise.ipaq.Constants
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.ClassPathResource
import org.springframework.data.redis.connection.RedisConnection
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.core.StringRedisTemplate
import org.springframework.data.redis.core.ValueOperations
import java.nio.charset.Charset

/**
 * Created by terrence on 26/5/17.
 */
@Configuration
class IntegrationConfig {

    @Bean
    fun connectionFactory(): RedisConnectionFactory {

        val redisConnectionFactory = mock(RedisConnectionFactory::class.java)
        val redisConnection = mock(RedisConnection::class.java)

        `when`(redisConnectionFactory.connection).thenReturn(redisConnection)

        return redisConnectionFactory
    }

    @Bean
    fun stringRedisTemplate(): StringRedisTemplate {

        val stringRedisTemplate = mock(StringRedisTemplate::class.java)
        val valueOperations = mock(ValueOperations::class.java)

        @Suppress("UNCHECKED_CAST")
        `when`(stringRedisTemplate.opsForValue()).thenReturn(valueOperations as ValueOperations<String, String>?)

        // Redis get()
        `when`(valueOperations!!.get(eq(Constants.QUERY_ADDRESS + ", " + Constants.QUERY_COUNTRY)))
                .thenReturn(StringUtils.EMPTY)
                .thenReturn(IOUtils.toString(ClassPathResource("search-resp.json").inputStream, Charset.defaultCharset()))

        `when`(valueOperations!!.get(eq(Constants.QUERY_ID + ", " + Constants.QUERY_COUNTRY)))
                .thenReturn(StringUtils.EMPTY)
                .thenReturn(IOUtils.toString(ClassPathResource("format-resp.json").inputStream, Charset.defaultCharset()))

        // Redis set()
        doNothing().`when`<ValueOperations<String, String>>(valueOperations).set(any(String::class.java), any(String::class.java))

        // Redis delete()
        doNothing().`when`(stringRedisTemplate).delete(anyString())

        return stringRedisTemplate
    }

}
